app/controlplane/internal/usercontext/apitoken_middleware.go

```diff
@@ -101,6 +102,98 @@ func WithCurrentAPITokenAndOrgMiddleware(apiTokenUC *biz.APITokenUseCase, orgUC
 	}
 }
 
+// WithCurrentAPITokenAndRobotAccountMiddleware injects the API-Token, organization + robot account to the context
+func WithCurrentAPITokenAndRobotAccountMiddleware(apiTokenUC *biz.APITokenUseCase, orgUC *biz.OrganizationUseCase, wfUC *biz.WorkflowUseCase, logger *log.Helper) middleware.Middleware {
+	return func(handler middleware.Handler) middleware.Handler {
+		return func(ctx context.Context, req interface{}) (interface{}, error) {
+			authInfo, ok := FromJWTAuthContext(ctx)
+			// If not found means that there is no currentUser set in the context
+			if !ok {
+				logger.Warn("couldn't extract org/user, JWT parser middleware not running before this one?")
+				return nil, errors.New("can't extract JWT info from the context")
+			}
+
+			// If the token is not an API token, we don't need to do anything
+			if !authInfo.IsAPIToken {
+				return handler(ctx, req)
+			}
+
+			genericClaims, ok := authInfo.Claims.(jwt.MapClaims)
+			if !ok {
+				return nil, errors.New("error mapping the claims")
+			}
+
+			// We've received an API-token
+			if genericClaims.VerifyAudience(apitoken.Audience, true) {
+				var err error
+				tokenID, ok := genericClaims["jti"].(string)
+				if !ok || tokenID == "" {
+					return nil, errors.New("error mapping the API-token claims")
+				}
+
+				ctx, err = setRobotAccountFromAPIToken(ctx, apiTokenUC, orgUC, wfUC, req, tokenID)
+				if err != nil {
+					return nil, fmt.Errorf("error setting current org and robot account: %w", err)
+				}
+
+				logger.Infow("msg", "[authN] processed credentials", "id", tokenID, "type", "API-token")
+			}
+
+			return handler(ctx, req)
+		}
+	}
+}
+
+func setRobotAccountFromAPIToken(ctx context.Context, apiTokenUC *biz.APITokenUseCase, orgUC *biz.OrganizationUseCase, wfUC *biz.WorkflowUseCase, req interface{}, tokenID string) (context.Context, error) {
+	if tokenID == "" {
+		return nil, errors.New("error retrieving the key ID from the API token")
+	}
+
+	// Check that the token exists and is not revoked
+	token, err := apiTokenUC.FindByID(ctx, tokenID)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving the API token: %w", err)
+	} else if token == nil {
+		return nil, errors.New("API token not found")
+	}
+
+	// Note: Expiration time does not need to be checked because that's done at the JWT
+	// verification layer, which happens before this middleware is called
+	if token.RevokedAt != nil {
+		return nil, errors.New("API token revoked")
+	}
+
+	// Find the associated organization
+	org, err := orgUC.FindByID(ctx, token.OrganizationID.String())
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving the organization: %w", err)
+	} else if org == nil {
+		return nil, errors.New("organization not found")
+	}
+
+	var workflowName string
```

Comment:
this might not be worth having here, I am wondering if we should just load the workflow in the service, it might be more transparent and easier to understand.

Reply:
If we move that information to the services, they will be the one in charge of from a workflowName, take the workflowID. This middleware here won't make much sense then because it will be only setting the `OrgID` since the token ID was recently removed.

Reply:
Moving it to the service means that service might receive "incomplete" requests. In the other side, doing database queries from the middlewares doesn't sound the right place. It would be nice if we figure out an intermediate solution, something like a common service initialization phase.

Reply:
If we decided to go through the middleware path, instead of trying to cast the `req interface{}` to each known request type, we could use reflection to check if the field exists (I don't see any other option).
```
message := req.(ProtoMessage)
message.ProtoReflect().Get(fieldDescriptor)
```

Reply:
TLDR

Personally, but take it with a grain of salt since I still need to review this patch thoroughly, I'd rather go to the approach where the services figure out how to get the workflow. And the middleware just a) `validate the API token JWT`, b) `checks that the token is not revoked` and c) `stores the org id in the robot-account (soon to be renamed) object` 

TLDR/

I see, to me, a rule of thumb is designing the solution from the POV of the API. I'm not too fond of the current approach (requiring name so the middleware can complete the stuff) because it forces us to make suboptimal API choices https://github.com/chainloop-dev/chainloop/pull/766

Knowing what I know now, I'd rather let the services figure out what's missing if needed, for example I can see that to get the workflowID

- att init (and contract get): let's take it from the provided workflow_name if needed 
- att push or get creds: get it from the workflowRunID

As you can see, there are different parameters in different API endpoints from where we can find the workflowID, so I think it's reasonable to let this logic be handled by the service layer, it's not generic enough.

> If we move that information to the services, they will be the one in charge of from a workflowName, take the workflowID. This middleware here won't make much sense then because it will be only setting the OrgID since the token ID was recently removed.

I think the middleware even in that case makes sene. It will do two things, validate that the token is not revoked and set the value of the OrgID. OrgID is very important to make sure we scope the calls to the biz layer.